### PR TITLE
PR with fixes of some issues

### LINKED ISSUE #487: Allow for arbitrary headers and do not force "; charset=...." in Content-Type
Current implementation of HttpListenerResponse.SendHeaders will always mess with Content-Type and add other headers. Although it may be convenient, this might be a problem in some cases.

Not all Content-Types support "charset" parameter. For instance,  "multipart/form-data" should have "boundry" instead. Also, sometimes there must not be Content-Type at all. Same thing with other default headers - there must be an option not to send them! 

Full control of headers should be doable from the request-handler code.  
Typical solution would be to expose Response.WriteHeaders() method:  once called explicitly from handler code, only current context.Response.Headers will be sent (which can be empty). This will not break the default SendHeaders() logic, since HeadersSent will already be set to true. That would allow user to have full control of headers when necessary while keeping the old behavior unchanged.

Currently there is no other workaround except modifying EmbedIO.Net.Internal implementation or or hacking-around with reflection.

### LINKED ISSUE #502: Overlapping WebSocket messages close the socket
**Describe the bug**
When client and EmbedIO heavily exchange messages on a WebSocket channel, the receiving and responding eventually overlap - especially with large messages. This results in the connection being closed, presumeably from the client (browser) side:

`WebSocket connection to 'ws://localhost:8080/event' failed: Received start of new message but previous message is unfinished.`

Have a look at [this sample project](https://github.com/bdurrer/embedio-websocket-example) simulating this behavior. It sends garbage text from and to the browser and waits randomly in the WebSocketModule to simulate actual work being done.

**To Reproduce**
1. Checkout the example project at https://github.com/bdurrer/embedio-websocket-example
1. Start the server (your editor might have a different output path than mine, check that the html-root folder path is correct in the logged output
1. Open http://localhost:8080/ in the browser
1. Open the browser console
1. Hit the _spam_ button
1. Wait until you get a red error shown in the console. It should take only seconds. (You can stop the spam by clicking the button again)

**Expected behavior**
Heavy usage of websockets should not make them fail. My understanding is, that websockets are capable of this and I do not have to create some complicated setup with locks for a simple ping-pong style websocket.
(If websockets weren't capable of handling this, it would mean they are a bad choice for chat software that has no chance predicting when and in what direction the next message travels).

I _might_ just have implemented the WebSocket incorrectly. If that's the case, I hope the outcome of this issue is a good example of a correct WebSocketModule 😊 

**Screenshots**
none

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome, Firefox
 - EmbedIO 3.4.3

**Smartphone (please complete the following information):**
 - does not apply

**Additional context**
I noticed this when attempting to convert a database-accessing api for a electron application, which uses a local EmbedIO server.
I mentioned this in the EmbedIO Slack on 7th of january. This is the correspoding bug report.

### LINKED ISSUE #504: HtmlDirectoryListener adds slash to the end of file URLs
**Description of the bug**
When listing files in directories using the HtmlDirectoryListener the URLs for the files end in a slash.
This causes issues with some tools such as wget which will download the file and save it as index.html instead of the name of the files.

**To Reproduce**
1. In a directory (/temp) put some files
2. Create a server using the following code
`
WebServer server = new WebServer(HttpListenerMode.EmbedIO, "http://*:8080");
server.WithLocalSessionManager();
server.WithModule(new EmbedIO.Files.FileModule("/", new EmbedIO.Files.FileSystemProvider(Path.GetFullPath(Path.Combine("/", "temp")), false))
{
    DirectoryLister = EmbedIO.Files.DirectoryLister.Html,
});
server.RunAsync();`
3. Open the servers web page in a web browser 
4. Examine the URLs for the files that are listed in the directory

**Expected behavior**
The URLs for listed files should not end with a slash.

**Screenshots**
Chrome browser from Windows
FileURLs
Chrome on Android
Screenshot_20210304-004122

**Desktop:**
 - OS: Windows 10
 - Browser chrome

**Smartphone:**
 - Device: Samsung Note5
 - OS: Android 7.0
 - Browser Chrome
 - Version ??

**Additional context**
Pull request: https://github.com/unosquare/embedio/pull/503

### LINKED ISSUE #519: Typo in README.md
Too small of an issue to create a PR, so I thought I'd just write it down here.

**Describe the bug**
In _README.md_, there is a _WebSockets Example_ provided.

The example class' names in this example are different. In first paragraph, it's "WebSocketChatModule", and in the next one, it is "WebSocketsChatServer" (that implements WebSocketModule, so I guess it's not really a "Server", but a "Module".)

### LINKED ISSUE #528: WebSocketModule.OnClientDisconnectedAsync called twice when closing and reopening websocket
### Context

Hello,
I am doing some tests with websockets. I close an existing websocket and I open a new one just after:
```javascript
if(ws)
   ws.close();
ws = new WebSocket(`ws://${host}/ws`, []);
```

I have an minimal WebsocketModule:
```csharp
    class TestWebSocketModule : WebSocketModule
    {
        public TestWebSocketModule(string urlPath) : base(urlPath, false) { }

        protected override Task OnMessageReceivedAsync(IWebSocketContext context, byte[] buffer, IWebSocketReceiveResult result)
        {
            string messageStr = Encoding.GetString(buffer);
            Debug.Log($"[{context.Id}] OnMessageReceivedAsync\n{messageStr}");
            return Task.CompletedTask;
        }

        protected override Task OnClientConnectedAsync(IWebSocketContext context)
        {
            Debug.Log($"[{context.Id}] OnClientConnectedAsync");
            return Task.CompletedTask;
        }

        protected override Task OnClientDisconnectedAsync(IWebSocketContext context)
        {
            Debug.Log($"[{context.Id}] OnClientDisconnectedAsync");
            return Task.CompletedTask;
        }
    }
```

### Observed Behaviour
The second time I re-open socket, `OnClientDisconnectedAsync` is called twice for the existing websocket that I juste closed.

I get these log messages:

> [189a7lnBhUe+ZWx0n6Dmtw] OnClientConnectedAsync (initial open)
> [ce14wIQRbkyfBgmzkNkQTA] OnClientConnectedAsync (1st repoen)
> [189a7lnBhUe+ZWx0n6Dmtw] OnClientDisconnectedAsync
> [ce14wIQRbkyfBgmzkNkQTA] OnClientDisconnectedAsync (2nd reopen)
> [STMccCfqYkq/83YctmrMbA] OnClientConnectedAsync
> [ce14wIQRbkyfBgmzkNkQTA] OnClientDisconnectedAsync

After tracing the code, It seems that `OnClientDisconnectedAsync` is called the first time from [`PurgeDisconnectedContexts()` in `OnRequestAsync()`](https://github.com/unosquare/embedio/blob/603c3deaf141fe7c79a609e02be106a2b3193c7f/src/EmbedIO/WebSockets/WebSocketModule.cs#L197) and the second time from [the `finally`block in `OnRequestAsync()`](https://github.com/unosquare/embedio/blob/603c3deaf141fe7c79a609e02be106a2b3193c7f/src/EmbedIO/WebSockets/WebSocketModule.cs#L231).

Enabling the watchdog or not does not change the behaviour

### Expected Behavior
I expect that `OnClientDisconnectedAsync` is only called once per context

### Versions
I observed the problem with EmbedIO v3.4.3 from nuget and build from source.
I tested with Firefox 89 and Chrome 91 on mac.
EmbedIO is called from the Unity Editor.

### LINKED ISSUE #531: Listener prefix parsing exception when using implicit default HTTP port.
**Listener prefix parsing issue.**
An exception is thrown when parsing a listener prefix that uses an implicit default HTTP port (80).

**To Reproduce**
1. Use "http://somehost/" as a listener prefix.
2. Use the prefix...
```csharp
EmbedIO.WebServer server = new EmbedIO.WebServer(
    configure => configure
       .WithUrlPrefix("http://somehost/")
       .WithMode(HttpListenerMode.EmbedIO)
)
```
3. See error
```
Unhandled Exception: L2 14:08:28.327273 (14016-26404) [AddListener] Length cannot be less than zero.
Parameter name: length; Length cannot be less than zero.
Parameter name: length    at System.String.Substring(Int32 startIndex, Int32 length)
   at EmbedIO.Net.Internal.ListenerPrefix..ctor(String uri) in D:\dev\embedio\src\EmbedIO\Net\Internal\ListenerPrefix.cs:line 29
   at EmbedIO.Net.EndPointManager.AddPrefix(String p, HttpListener listener) in D:\dev\embedio\src\EmbedIO\Net\EndPointManager.cs:line 75
System.ArgumentOutOfRangeException: Length cannot be less than zero.
Parameter name: length
ndPointManager.AddListener(HttpListener listener) in D:\dev\embedio\src\EmbedIO\Net\EndPointManager.cs:line 48
   at System.String.Substring(Int32 startIndex, Int32 length)
   at EmbedIO.Net.HttpListener.Start() in D:\dev\embedio\src\EmbedIO\Net\HttpListener.cs:line 65
   at EmbedIO.Net.Internal.ListenerPrefix..ctor(String uri) in D:\dev\embedio\src\EmbedIO\Net\Internal\ListenerPrefix.cs:line 29
   at EmbedIO.Net.EndPointManager.AddPrefix(String p, HttpListener listener) in D:\dev\embedio\src\EmbedIO\Net\EndPointManager.cs:line 75
ken cancellationToken) in D:\dev\embedio\src\EmbedIO\WebServer.cs:line 140
   at EmbedIO.Net.EndPointManager.AddListener(HttpListener listener) in D:\dev\embedio\src\EmbedIO\Net\EndPointManager.cs:line 48
   at EmbedIO.WebServerBase`1.<RunAsync>d__39.MoveNext() in D:\dev\embedio\src\EmbedIO\WebServerBase`1.cs:line 198
   at EmbedIO.Net.HttpListener.Start() in D:\dev\embedio\src\EmbedIO\Net\HttpListener.cs:line 65
```

**Expected behavior**
Parse the listener prefix without an exception and use the default port 80 as the port number.

**Additional context**
The error seems to be related to the line:
```csharp
var colon = uri.LastIndexOf(':');
```
If there is no port-related colon, this will find the index of the colon after the URI scheme ("http:"). The later (port-related) colon index check,
```csharp
if (colon > 0)
{
    Host = uri.Substring(startHost, colon - startHost);
    // . . .
```
will incorrectly pass and in the next line the computed Substring index will be negative and trigger the exception.

### PR DESCRIPTION
**Pull request by radioegor146**
Fixes #531 
Fixes #528
Fixes #519 
Fixes #504 
Fixes #502 
Fixes #487
Fixes tests on GH Actions

### CODE REVIEW

**Comment by radioegor146** [comment]
@rdeago also, is there any possible time of release v4.0? Or I can continue to fix that library from the current state?

File: README.md [hunk 1]
```diff
@@ -223,9 +223,9 @@ namespace Unosquare
     /// <summary>
     /// Defines a very simple chat server.
     /// </summary>
-    public class WebSocketsChatServer : WebSocketModule
+    public class WebSocketsChatModule : WebSocketModule
```
Comment (rdeago) [inline README.md:226]: Not really addressing an issue, but good for consistency anyway. 👍

File: src/packages/EmbedIO/Files/ZipFileProvider.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ public void Start(CancellationToken cancellationToken)
 
             path = Uri.UnescapeDataString(path);
 
-            var entry = _zipArchive.GetEntry(path.Substring(1));
+            ZipArchiveEntry entry = _zipArchive.GetEntry(path.Substring(1));
```
Comment (rdeago) [inline src/packages/EmbedIO/Files/ZipFileProvider.cs:79]: Why not just use `var`?

File: src/packages/EmbedIO/Net/Internal/SystemHttpResponse.cs [hunk 3]
```diff
@@ -1,6 +1,8 @@
 ﻿using System;
+using System.Collections;
```
Comment (rdeago) [inline src/packages/EmbedIO/Net/Internal/SystemHttpResponse.cs:2]: Are you sure these two added usings are necessary? You made no other changes to this file.

**Review by rdeago — CHANGES_REQUESTED** [review]
Great job @radioegor146! Please forgive me for taking so long to review.

Being a fan of simplification, I especially like `ListenerUri`. 😃 

Just remove the two added usings [here](https://github.com/unosquare/embedio/pull/534/files#diff-d050adcf0087ebd5cb8f9ce888f59375bd2c70f516a4eeb37da6edf8971dddb1) if they are not necessary. Otherwise, this PR is absolutely to be merged IMO.

**Comment by radioegor146** [comment]
> Great job @radioegor146! Please forgive me for taking so long to review.
> 
> Being a fan of simplification, I especially like `ListenerUri`. 😃
> 
> Just remove the two added usings [here](https://github.com/unosquare/embedio/pull/534/files#diff-d050adcf0087ebd5cb8f9ce888f59375bd2c70f516a4eeb37da6edf8971dddb1) if they are not necessary. Otherwise, this PR is absolutely to be merged IMO.

I've fixed some changes. Also, the added `chcp` command is to make sure that `netsh` output is in the simplest possible encoding (CP437/OEM), just because on systems with other default regional settings (as mine - ru_RU), the error message is really unreadable.

File: src/packages/EmbedIO/Files/ZipFileProvider.cs (hunk 2 above)
Comment (rdeago) [reply src/packages/EmbedIO/Files/ZipFileProvider.cs]: Solved in later commit. 👍

File: src/packages/EmbedIO/Net/Internal/SystemHttpResponse.cs (hunk 3 above)
Comment (rdeago) [reply src/packages/EmbedIO/Net/Internal/SystemHttpResponse.cs]: Solved in later commit. 👍

**Review by rdeago — APPROVED** [review]
Great job @radioegor146!

**Comment by rdeago** [comment]
> @rdeago also, are there any plans or possible time of release v4.0? Or I can continue to fix that library/add new features from the current state?

Unfortunately, no 4.0 release in sight at the moment, but I'm going to submit a PR that backports your changes to the `v3.x` branch.

**Comment by dtsirkovmb** [comment]
Hi @rdeago, do you have any idea when will backports to v3.x take place? Cheers.

**Comment by rdeago** [comment]
@dtsirkovmb I'm on it. I also found a couple other things in need of a fix, the PR should be ready by Monday.

**Pull request merged**